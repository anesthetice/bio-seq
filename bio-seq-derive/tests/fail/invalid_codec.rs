use bio_seq::prelude::*;

#[derive(Codec)]
struct NotEnum {
    a: (),
    c: (),
    g: (),
    t: (),
}

#[derive(Codec)]
enum NoDiscriminants {
    A,
    C,
    G,
    T,
}

#[derive(Codec, PartialEq, Debug, Hash)]
#[bits(2)]
enum BadWidth {
    A = 0b000,
    C = 0b001,
    G = 0b010,
    T = 0b100,
    X = 0b011,
}

#[derive(Codec, PartialEq, Debug, Hash, Eq, Copy, Clone)]
#[bits(3)]
enum BadDiscriminant {
    A = 0b000,
    C = 3,
    G = 0b010,
    T = 0x04,
    X = 1.0,
}

// Test whether codec is #[repr(u8)]
/*
#[derive(Codec, Eq, Debug, Hash, PartialEq, Copy, Clone)]
enum NoRepr {
    A = 0b11,
    C = 0b01,
    G = 0b00,
    T = 0b10,
}
*/

fn main() {
    let _c = NotEnum {
        a: (),
        c: (),
        g: (),
        t: (),
    };
}
