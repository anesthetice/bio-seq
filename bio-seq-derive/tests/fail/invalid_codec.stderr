error: Codec can only be derived for enums
 --> tests/fail/invalid_codec.rs:4:1
  |
4 | / struct NotEnum {
5 | |     a: (),
6 | |     c: (),
7 | |     g: (),
8 | |     t: (),
9 | | }
  | |_^

error: Codec derivations require discriminants
  --> tests/fail/invalid_codec.rs:13:5
   |
13 |     A,
   |     ^

error: Bit width is not large enough encode all variants (min: 3)
  --> tests/fail/invalid_codec.rs:20:1
   |
20 | #[bits(2)]
   | ^^^^^^^^^^

error: Codec derivations require byte or integer discriminants
  --> tests/fail/invalid_codec.rs:36:5
   |
36 |     X = 1.0,
   |     ^

error[E0308]: mismatched types
  --> tests/fail/invalid_codec.rs:36:9
   |
36 |     X = 1.0,
   |         ^^^ expected `isize`, found floating-point number
